cortex/utils/commands.py

```
@@ -0,0 +1,344 @@
+                ]
+                # Check if all $(...) patterns are in safe list
+                found_subs = re.findall(r'\$\([^)]+\)', command)
+                for sub in found_subs:
+                    if sub not in safe_substitutions:
+                        return False, f"Unsafe command substitution: {sub}"
+            elif char == '&&' or char == '||':
+                # Allow chained commands, but validate each part
+                continue
+            elif char == ';':
+                # Semicolon is dangerous - could chain arbitrary commands
+                return False, f"Semicolon not allowed in commands"
```

Ruff (0.14.6)

f-string without any placeholders

Remove extraneous `f` prefix

(F541)